pgspecial/dbcommands.py

```diff
@@ -1524,3 +1526,9 @@ def show_function_definition(cur, pattern, verbose):
     else:
         headers = None
     return [(None, cur, headers, None)]
+
+@special_command('\\!', '\\![+] [commands]', 'Pass commands to shell.')
+def shell_command(cur, pattern, verbose):
+    cur, headers = [], []
+    params = shlex.split(pattern)
+    return [(None, cur, headers, subprocess.run([param for param in params]).stdout)]
```

Comment:
Why not `subprocess.run(params)`?

Reply:
Also, I believe that `subprocess.run` will not capture stdout by default. You have to provide `stdout=PIPE`. This is off the top of my head, definitely test that part.

Reply:
> Why not `subprocess.run(params)`?

No reason not to. You're right.

> Also, I believe that `subprocess.run` will not capture stdout by default.

It doesn't, however it's being called by the `.stdout` property

Reply:
I forgot to add that `subprocess.run()` was introduced in Python 3.5

https://docs.python.org/3/library/subprocess.html#subprocess.run

Reply:
Whatever works. This part, however, does not look great in `call` documentation:

```
Note Do not use stdout=PIPE or stderr=PIPE with this function as that can deadlock based on the child process output volume. Use Popen with the communicate() method when you need pipes.
```

Reply:
I believe that note in the documentation refers to explicitly setting the `stdout` argument to `PIPE`, as in `subprocess.call(*, stdout=PIPE)`. The default value is `None` so there should be no issues with that.

I did tests in 2.7, 3.6 and 3.7 passing both `\! pwd` and `\! ls -l` and it worked well.

Reply:
Interesting. And was `output` not `None`?

Reply:
Actually, you won't get output from `call`, right? only return code?

Reply:
I may be misunderstanding something. I did not run this code, only read it. I am assuming that you want to:
* run a shell command
* capture whatever text output
* display it within pgcli as `status`
My problem with s code is that I think `subprocess.call` will not capture that output. Am I mistaken?

Reply:
No worries. You've got the goal correctly. `subprocess.call` does capture the output. As of right now it literally works just like running the commands directly on a shell 😎

I made a terrible gif [over here](https://imgur.com/a/SZZ22BG). That one specifically is on 2.7

Reply:
Ok, I see how it works. It does not work because output is captured into variable, but because it is passed through to console. Example:

```
Python 2.7.14 (default, Apr  5 2018, 18:02:59)
>>> import subprocess
>>> result = subprocess.call(['ls', '-l'])
total 240
-rw-r--r--   1 irina  staff   1718 Nov 18 16:32 AUTHORS
-rw-r--r--   1 irina  staff   5730 Nov 13 14:02 DEVELOP.rst
>>> print(result)
0
```
so in fact you'll get `0` as your `status`.